### ISSUE TITLE
Transmission-3e958cfbaf Build #9545 crash

### ISSUE THREAD

**Issue by dantearmok** [transmission/transmission#6744]
### What is the issue?

Mac mini (late 2012, 2.5-GHz 4-Core i5), 16GB RAM, plus an assortment of USB & FW HDs.
macOS 14.4.1 Sonoma (w/OCLP 1.4.2), PIA 3.5.7, and Xcode.
Pipe is 450 Mbps wifi to Comcast coax 100/25 Mbps, with PIA beaming me thru to Canada.

Running Transmission-3e958cfbaf Build #9545 (Mar 25, 2024, 1:51:30 AM)

With 1,006 torrents running, Transmission was doing well.  Then it quietly crashed...

-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               Transmission [2931]
Path:                  /Users/USER/*/Transmission-2ff3ae07d1 Build #9551.app/Contents/MacOS/Transmission
Identifier:            org.m0k.transmission
Version:               4.1.0-dev (14718.1.0)
Code Type:             X86-64 (Native)
Parent Process:        launchd [1]
User ID:               501

Date/Time:             2024-03-27 15:21:10.0748 -0400
OS Version:            macOS 14.4.1 (23E224)
Report Version:        12
Anonymous UUID:        C5048D24-7417-7AA9-D745-755E4ED63C56

Time Awake Since Boot: 4200 seconds

System Integrity Protection: enabled

Crashed Thread:        1

Exception Type:        EXC_CRASH (SIGABRT)
Exception Codes:       0x0000000000000000, 0x0000000000000000

Termination Reason:    Namespace SIGNAL, Code 6 Abort trap: 6
Terminating Process:   Transmission [2931]

Application Specific Information:
abort() called

Thread 0::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	    0x7ff8084af3ce mach_msg2_trap + 10
1   libsystem_kernel.dylib        	    0x7ff8084bdc88 mach_msg2_internal + 84
2   libsystem_kernel.dylib        	    0x7ff8084b6510 mach_msg_overwrite + 653
3   libsystem_kernel.dylib        	    0x7ff8084af6bd mach_msg + 19
4   CoreFoundation                	    0x7ff8085cc359 __CFRunLoopServiceMachPort + 143
5   CoreFoundation                	    0x7ff8085cadcc __CFRunLoopRun + 1371
6   CoreFoundation                	    0x7ff8085ca2a9 CFRunLoopRunSpecific + 557
7   HIToolbox                     	    0x7ff8134b5829 RunCurrentEventLoopInMode + 292
8   HIToolbox                     	    0x7ff8134b5636 ReceiveNextEventCommon + 665
9   HIToolbox                     	    0x7ff8134b5381 _BlockUntilNextEventMatchingListInModeWithFilter + 66
10  AppKit                        	    0x7ff80bb93be5 _DPSNextEvent + 880
11  AppKit                        	    0x7ff80c4a3fe9 -[NSApplication(NSEventRouting) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1273
12  AppKit                        	    0x7ff80bb85005 -[NSApplication run] + 603
13  AppKit                        	    0x7ff80bb58ff1 NSApplicationMain + 816
14  Transmission                  	       0x10e73307f main + 73 (main.mm:17)
15  dyld                          	    0x7ff808163366 start + 1942

Thread 1 Crashed:
0   libsystem_kernel.dylib        	    0x7ff8084b614a __pthread_kill + 10
1   libsystem_pthread.dylib       	    0x7ff8084eeebd pthread_kill + 262
2   libsystem_c.dylib             	    0x7ff808414a39 abort + 126
3   Transmission                  	       0x10e8651b7 tr_assert_report(std::__1::basic_string_view<char, std::__1::char_traits<char>>, long, std::__1::basic_string_view<char, std::__1::char_traits<char>>) + 135 (tr-assert.cc:16)
4   Transmission                  	       0x10e8273e6 tr_netOpenPeerSocket(tr_session*, tr_socket_address const&, bool) + 4123 (net.cc:231)
5   Transmission                  	       0x10e842c10 tr_peerIo::reconnect() + 140 (peer-io.cc:248)
6   Transmission                  	       0x10e83f57a tr_handshake::on_error(tr_peerIo*, tr_error const&, void*) + 266 (handshake.cc:654)
7   Transmission                  	       0x10e8445a5 tr_peerIo::call_error_callback(tr_error const&) + 31 (peer-io.h:340) [inlined]
8   Transmission                  	       0x10e8445a5 tr_peerIo::on_utp_error(int) + 621 (peer-io.cc:703)
9   Transmission                  	       0x10e84725a tr_peerIo::utp_init(struct_utp_context*)::$_5::operator()(utp_callback_arguments*) const + 25 (peer-io.cc:759) [inlined]
10  Transmission                  	       0x10e84725a tr_peerIo::utp_init(struct_utp_context*)::$_5::__invoke(utp_callback_arguments*) + 34 (peer-io.cc:755)
11  Transmission                  	       0x10e90f3c6 utp_call_on_error(struct_utp_context*, UTPSocket*, int) + 43 (utp_callbacks.cpp:69)
12  Transmission                  	       0x10e914872 auto utp_check_timeouts::$_0::operator()<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const>(std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) const + 14 (utp_internal.cpp:3283) [inlined]
13  Transmission                  	       0x10e914872 decltype(static_cast<utp_check_timeouts::$_0&>(fp)(static_cast<std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&>(fp0))) std::__1::__invoke<utp_check_timeouts::$_0&, std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&>(utp_check_timeouts::$_0&, std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) + 14 (type_traits:3918) [inlined]
14  Transmission                  	       0x10e914872 bool std::__1::__invoke_void_return_wrapper<bool, false>::__call<utp_check_timeouts::$_0&, std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&>(utp_check_timeouts::$_0&, std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) + 14 (invoke.h:30) [inlined]
15  Transmission                  	       0x10e914872 std::__1::__function::__alloc_func<utp_check_timeouts::$_0, std::__1::allocator<utp_check_timeouts::$_0>, bool (std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&)>::operator()(std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) + 14 (function.h:178) [inlined]
16  Transmission                  	       0x10e914872 std::__1::__function::__func<utp_check_timeouts::$_0, std::__1::allocator<utp_check_timeouts::$_0>, bool (std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&)>::operator()(std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) + 18 (function.h:352)
17  Transmission                  	       0x10e913714 std::__1::__function::__value_func<bool (std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&)>::operator()(std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) const + 18 (function.h:505) [inlined]
18  Transmission                  	       0x10e913714 std::__1::function<bool (std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&)>::operator()(std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&) const + 18 (function.h:1182) [inlined]
19  Transmission                  	       0x10e913714 UTPSocketHT::erase_if(std::__1::function<bool (std::__1::__hash_map_iterator<std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<UTPSocketKey, std::__1::unique_ptr<UTPSocket, UTPSocketHT::SocketDeleter>>, void*>*>> const&)>) + 62 (utp_internal.h:149)
20  Transmission                  	       0x10e913545 utp_check_timeouts + 356 (utp_internal.cpp:3280)
21  Transmission                  	       0x10e8b7dc4 (anonymous namespace)::timer_callback(void*) + 21 (tr-utp.cc:176)
22  Transmission                  	       0x10e8f7a00 event_process_active_single_queue + 502 (event.c:1691)
23  Transmission                  	       0x10e8f4b0b event_process_active + 220 [inlined]
24  Transmission                  	       0x10e8f4b0b event_base_loop + 1134 (event.c:2006)
25  Transmission                  	       0x10e833f75 tr_session_thread_impl::session_thread_func(event_base*) + 87 (session-thread.cc:244)
26  Transmission                  	       0x10e83435f decltype(*static_cast<tr_session_thread_impl*>(fp0).*fp(static_cast<event_base*>(fp1))) std::__1::__invoke<void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*, void>(void (tr_session_thread_impl::*&&)(event_base*), tr_session_thread_impl*&&, event_base*&&) + 30 (type_traits:3859) [inlined]
27  Transmission                  	       0x10e83435f void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*, 2ul, 3ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*>&, std::__1::__tuple_indices<2ul, 3ul>) + 30 (thread:287) [inlined]
28  Transmission                  	       0x10e83435f void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_session_thread_impl::*)(event_base*), tr_session_thread_impl*, event_base*>>(void*) + 63 (thread:298)
29  libsystem_pthread.dylib       	    0x7ff8084ef18b _pthread_start + 99
30  libsystem_pthread.dylib       	    0x7ff8084eaae3 thread_start + 15

Thread 2:
0   libsystem_kernel.dylib        	    0x7ff8084b1e46 __semwait_signal + 10
1   libsystem_c.dylib             	    0x7ff8083a4531 nanosleep + 199
2   libc++.1.dylib                	    0x7ff8084390b4 std::__1::this_thread::sleep_for(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l>> const&) + 73
3   Transmission                  	       0x10e8851b2 void std::__1::this_thread::sleep_for<long long, std::__1::ratio<1l, 1000l>>(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l>> const&) + 17 (thread:386) [inlined]
4   Transmission                  	       0x10e8851b2 tr_web::Impl::curlThreadFunc() + 448 (web.cc:717)
5   Transmission                  	       0x10e88544b decltype(*static_cast<tr_web::Impl*>(fp0).*fp()) std::__1::__invoke<void (tr_web::Impl::*)(), tr_web::Impl*, void>(void (tr_web::Impl::*&&)(), tr_web::Impl*&&) + 26 (type_traits:3859) [inlined]
6   Transmission                  	       0x10e88544b void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_web::Impl::*)(), tr_web::Impl*, 2ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_web::Impl::*)(), tr_web::Impl*>&, std::__1::__tuple_indices<2ul>) + 26 (thread:287) [inlined]
7   Transmission                  	       0x10e88544b void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void (tr_web::Impl::*)(), tr_web::Impl*>>(void*) + 59 (thread:298)
8   libsystem_pthread.dylib       	    0x7ff8084ef18b _pthread_start + 99
9   libsystem_pthread.dylib       	    0x7ff8084eaae3 thread_start + 15

Thread 3:: VDKQueue
0   libsystem_kernel.dylib        	    0x7ff8084b40ae kevent + 10
1   Transmission                  	       0x10e7db3e8 -[VDKQueue watcherThread:] + 240 (VDKQueue.mm:183)
2   Foundation                    	    0x7ff8094d63cc __NSThread__start__ + 1013
3   libsystem_pthread.dylib       	    0x7ff8084ef18b _pthread_start + 99
4   libsystem_pthread.dylib       	    0x7ff8084eaae3 thread_start + 15

Thread 4:: com.apple.NSEventThread
0   libsystem_kernel.dylib        	    0x7ff8084af3ce mach_msg2_trap + 10
1   libsystem_kernel.dylib        	    0x7ff8084bdc88 mach_msg2_internal + 84
2   libsystem_kernel.dylib        	    0x7ff8084b6510 mach_msg_overwrite + 653
3   libsystem_kernel.dylib        	    0x7ff8084af6bd mach_msg + 19
4   CoreFoundation                	    0x7ff8085cc359 __CFRunLoopServiceMachPort + 143
5   CoreFoundation                	    0x7ff8085cadcc __CFRunLoopRun + 1371
6   CoreFoundation                	    0x7ff8085ca2a9 CFRunLoopRunSpecific + 557
7   AppKit                        	    0x7ff80bcf0abc _NSEventThread + 122
8   libsystem_pthread.dylib       	    0x7ff8084ef18b _pthread_start + 99
9   libsystem_pthread.dylib       	    0x7ff8084eaae3 thread_start + 15

Thread 5:
0   libsystem_pthread.dylib       	    0x7ff8084eaac0 start_wqthread + 0

Thread 6:
0   libsystem_pthread.dylib       	    0x7ff8084eaac0 start_wqthread + 0

Thread 7:
0   libsystem_pthread.dylib       	    0x7ff8084eaac0 start_wqthread + 0

Thread 8:
0   libsystem_pthread.dylib       	    0x7ff8084eaac0 start_wqthread + 0

Thread 1 crashed with X86 Thread State (64-bit):
  rax: 0x0000000000000000  rbx: 0x0000000000000006  rcx: 0x000070000bd651b8  rdx: 0x0000000000000000
  rdi: 0x0000000000007703  rsi: 0x0000000000000006  rbp: 0x000070000bd651e0  rsp: 0x000070000bd651b8
   r8: 0x00007ff84b8af810   r9: 0x00000000ffffff00  r10: 0x0000000000000000  r11: 0x0000000000000246
  r12: 0x0000000000007703  r13: 0x000000010e9509b9  r14: 0x000070000bd66000  r15: 0x0000000000000016
  rip: 0x00007ff8084b614a  rfl: 0x0000000000000246  cr2: 0x0000000000000000
  
Logical CPU:     0
Error Code:      0x02000148 
Trap Number:     133

### Which application of Transmission?

macOS app

### Which version of Transmission?

Transmission-3e958cfbaf Build #9545 (Mar 25, 2024, 1:51:30 AM)

**Comment by tearfur** [comment]
**tl;dr** I can only assume that this was another invalid memory write corrupting data structures. Do you think you can compile Transmission from source with `fsanitize` enabled and try to catch where it is?

The crash caused by a failed assert caused by the `tr_peer_socket` object containing an invalid address when trying to reconnct to a timed out former uTP peer via TCP:

https://github.com/transmission/transmission/blob/3e958cfbaf29001e67f50a7595f1e621274d35fa/libtransmission/net.cc#L231

https://github.com/transmission/transmission/blob/3e958cfbaf29001e67f50a7595f1e621274d35fa/libtransmission/peer-io.cc#L248

Which shouldn't happen, because there are checks in place to prevent such a connect to be created in the first place.

- In the incoming case, the check is here:
   
   https://github.com/transmission/transmission/blob/3e958cfbaf29001e67f50a7595f1e621274d35fa/libtransmission/tr-utp.cc#L93-L96

- For the outgoing case, there is an assert that should have caught the invalid address:
   
   https://github.com/transmission/transmission/blob/3e958cfbaf29001e67f50a7595f1e621274d35fa/libtransmission/peer-io.cc#L129

And, the socket address is never modified once the `tr_peer_socket` is constructed.

So, the socket address had to have been corrupted after it was created.

**Comment by Coeur** [comment]
I reproduced immediately on launch.
>assertion failed: addr.is_valid() (/Users/coeur/Developer/transmission/libtransmission/net.cc:231)

![Capture d’écran 2024-03-28 à 22 55 38](https://github.com/transmission/transmission/assets/839992/cdfad819-89df-4f03-9f8b-5dc8b3b72d93)

The culprit line:
```
socket_ = tr_netOpenPeerSocket(session_, socket_address(), is_seed());
```
This created a new address `socket_address()` which is invalid.

**Comment by tearfur** [comment]
I see, this has been a bug for a long time. but it was being swept under the rug before #6717, because `tr_address` objects were labeled as IPv4 by default...

**Comment by tearfur** [comment]
No, it was not a regression. #6717 merely uncovered a long existing bug.

**Comment by Coeur** [comment]
It wasn't crashing before that commit, that's what I define as "regression", even if there was already another bug.

**Comment by tearfur** [comment]
And because #6717 classifies as a "regression" to you, reverting it became an option? I'm not convinced that logic is very sound.

In my eyes #6717 set a couple of things straight, and thanks to it we discover that we have been trying to reconnect to invalid socket addresses, and the asserts that were supposed to catch it was not doing its job. I disagree with reverting it.

So far I see 2 code smells that needs fixing:
- Intuitively speaking, it doesn't make sense that failing to reconnect immediately discards the socket address of the peer. It's entirely valid to keep it for cleanup and bookkeeping. So callers of `tr_peerIo::reconnect()` shouldn't have to worry about that caveat in the first place.
- The symtoms for "peer does not support uTP" and "peer does not support encryption" are *very* similar, to the point that we might not have enough information to identify them, and that's why `tr_peerIo::reconnect()` ended up being called twice. If you think about it, why should it *ever* be called more than once for each endpoint? It failed the first time, 99% chance it will also fail the second time.

What we should do is figure out how to restructure the code to properly address these code smells, instead of just adding more and more `if` conditions to mask the problem. It'll just come back to bite us later.

**Comment by Coeur** [comment]
I don't mean to avoid a clean fix: I mean that major regressions should just be _immediately_ reverted, and _then_ we spend time for finding the right approach, but without impacting the users or developers with major problems.

**Comment by nevack** [comment]
> I don't mean to avoid a clean fix: I mean that major regressions should just be _immediately_ reverted, and _then_ we spend time for finding the right approach, but without impacting the users or developers with major problems.

How is it a major regression? I thought `main` branch is for development and anybody using nightlies must be ready for such thing.

If this issue getting in the way for you - rebase your branches onto last known good revision.

I'm on @tearfur side on this one - we can afford figuring this out properly, even if some time is needed for it.
